# `empty-file`

> Category: style
>
> Enabled by default?: No

## What This Rule Does

Explain what this rule checks for. Also explain why this is a problem.

## Examples

Examples of **incorrect** code for this rule:

```zig

```

Examples of **correct** code for this rule:

```zig
fn exampleFunction() void {
}
```
