---
name: Rule Suggestion
about: Suggest a new lint rule
title: 'rule: `some-rule-name`'
labels: A-linter, C-rule
assignees: ''

---

## Description
Please describe what syntax or coding patterns this rule prohibits. Make sure you describe _why_ its bad.

## Examples

Examples of **incorrect** code for this rule
```zig
// put examples of code that should produce any violations
```

Examples of **correct** code for this rule
```zig
// Put examples of code that should _not_ produce any violations
```
