# `empty-file`

<RuleBanner category="style" default="warning"  />

## What This Rule Does
This rule checks for empty .zig files in the project.
A file should be deemed empty if it has no content (zero bytes) or only comments and whitespace characters,
as defined by the standard library in [`std.ascii.whitespace`](https://ziglang.org/documentation/master/std/#std.ascii.whitespace).

## Examples

Examples of **incorrect** code for this rule:


```zig
// an "empty" file is actually a file without meaningful code: just comments (doc or normal) or whitespace
```

Examples of **correct** code for this rule:
```zig
fn exampleFunction() void {
}
```

## Configuration
This rule has no configuration.
